akka-http/src/main/scala/akka/http/scaladsl/server/directives/AroundRequestDirectives.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.scaladsl.server
+package directives
+
+import akka.NotUsed
+import akka.http.scaladsl.server.RouteResult.{ Complete, Rejected }
+import akka.http.scaladsl.util.FastFuture._
+import akka.stream.scaladsl.Flow
+import akka.stream.{ FlowShape, Graph }
+import akka.util.ByteString
+
+import scala.concurrent.{ ExecutionContextExecutor, Future }
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * @groupname around AroundRequest directives
+ * TODO: group priority??
+ * @groupprio around 240
+ */
+trait AroundRequestDirectives {
+
+  import TimeoutDirectives.withRequestTimeoutOnDone
+
+  /**
+   * Partially applies supplied function on new requests.
+   * The resulting new function will be applied when inner route completes, rejects, or throws an error.
+   *
+   * If entity is present in response, onDone will be invoked on last byte sent.
+   *
+   * Attempts to apply new function on timeouts. On timeouts, function is applied synchronously and shouldn't block!
+   * Due to the inherent raciness it is not guaranteed that the update will be applied before
+   * the previously set timeout has expired!
+   *
+   * @group around
+   */
+
+  def aroundRequest(onRequest: RequestContext => Try[RouteResult] => Unit): Directive0 =
+    Directive { inner => ctx =>
+      import ctx.executionContext
+      val onDone = onRequest(ctx)
+      withRequestTimeoutOnDone(r => {
+        onDone(Success(Complete(r)))
+        r
+      })(inner(()))(ctx).fast.map {
+        case c: Complete if c.response.entity.isKnownEmpty => onDone(Success(c)); c
+        case c: Complete                                   => Complete(c.response.transformEntityDataBytes(onEntityTermination(onDone, c)))
+        case r: Rejected                                   => onDone(Success(r)); r
+      }.recoverWith({
+        case e: Throwable =>
+          onDone(Failure(e))
+          Future.failed(e)
+      })
+    }
+
+  private def onEntityTermination(onDone: Try[RouteResult] => Unit, result: RouteResult)(implicit ec: ExecutionContextExecutor): Graph[FlowShape[ByteString, ByteString], NotUsed] =
+    Flow[ByteString].watchTermination() {
```

Comment:
`watchTermination` is built on a GraphStage, which only supports `onUpstreamFinish()`, and it's `onDownstreamFinish()` is only called when there's an exception. See https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala#L128.

In case of the chunk response, it'll stop before the last chunk is streamed to the client, and in case of a large response, it'll be the first byte time.

See https://github.com/ssgao/akka-http-quickstart-scala for simple demo.

Reply:
I'm not seeing this issue. Hopefully I'm not missing something. 

I tested this 2 ways.

1. This unit test: https://github.com/akka/akka-http/pull/2814/files#diff-5a8c270ba581eaaa8c1ad33be1e3202eR134

2. I cloned the repo you linked and tested with a large file - `1.3GB`

**Procedure:**
Configure curl to measure transfer times. Create a file

Reply:
I tried the second route named `fooc` in the example app and got similar results

```
tom.molloy@...:~/repos/public/akka-http-quickstart-scala (master)$ curl -w "@curl-format.txt" -o /dev/null -s "http://localhost:8080/fooc"
time_namelookup:  0.005077
time_connect:  0.005419
time_appconnect:  0.000000
time_pretransfer:  0.005470
time_redirect:  0.000000
time_starttransfer:  0.294510
----------
time_total:  18.760435
```

```
root [helloAkkaHttpServer-akka.actor.default-dispatcher-3] INFO  c.e.QuickstartServer$ - [200] GET http://localhost:8080/fooc took: 18647ms
```

Reply:
> `watchTermination` is built on a GraphStage, which only supports `onUpstreamFinish()`, and it's `onDownstreamFinish()` is only called when there's an exception. 
Can some transparent Flow in the bottom send Cancel/Done/NonFailureCancellation etc up to signal when it receive its onUpstreamFinish? So we can ensure that streams are drain?

Reply:
Interesting idea. This might be a simple solution if it's doable.

Another idea I was thinking of is adding a function in `OutgoingConnectionBlueprint.scala`. The function will be applied when `MessageEnd` is received.

We can add a synthetic header to the request which allows access and modification of the function similar to how `Timeout-Access` is implemented.